### ISSUE TITLE
Slow API (TLS-Handshake)

### ISSUE THREAD

**Issue by Allandor** [Icinga/icinga2#6711]
The problem occured first after the upgrade to Version 2.10.0. 

Icingaweb is unable to connect to the API for "Command Transport"
```
icinga2: Operation timed out after 30001 milliseconds with 0 out of 0 bytes received.

#0 /usr/share/icingaweb2/modules/monitoring/application/forms/Command/Object/CheckNowCommandForm.php(70): Icinga\Module\Monitoring\Command\Transport\CommandTransport->send(Object(Icinga\Module\Monitoring\Command\Object\ScheduleServiceCheckCommand))
#1 /usr/share/php/Icinga/Web/Form.php(1158): Icinga\Module\Monitoring\Forms\Command\Object\CheckNowCommandForm->onSuccess()
#2 /usr/share/icingaweb2/modules/monitoring/library/Monitoring/Web/Controller/MonitoredObjectController.php(275): Icinga\Web\Form->handleRequest()
#3 /usr/share/icingaweb2/modules/monitoring/library/Monitoring/Web/Controller/MonitoredObjectController.php(60): Icinga\Module\Monitoring\Web\Controller\MonitoredObjectController->setupQuickActionForms()
#4 /usr/share/icingaweb2/modules/monitoring/application/controllers/ServiceController.php(66): Icinga\Module\Monitoring\Web\Controller\MonitoredObjectController->showAction()
#5 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Monitoring\Controllers\ServiceController->showAction()
#6 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch(String)
#7 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#8 /usr/share/php/Icinga/Application/Web.php(300): Zend_Controller_Front->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#9 /usr/share/php/Icinga/Application/webrouter.php(104): Icinga\Application\Web->dispatch()
#10 /usr/share/icingaweb2/public/index.php(4): require_once(String)
#11 {main}
```
Also other API-Clients are having trouble connecting to the API, but can establish a connection after around 2 minutes.
The Client trys to reschedule all checks for the Host.
The TLS handshake hangs on the Server hello message.
```
* Hostname was NOT found in DNS cache
*   Trying <redacted>...
* Connected to icinga.<redacted>.de (<redacted>) port 5665 (#0)
* successfully set certificate verify locations:
*   CAfile: /var/lib/icinga2/certs/ca.crt
  CApath: /etc/ssl/certs/
* SSLv3, TLS unknown, Certificate Status (22):
* SSLv3, TLS handshake, Client hello (1):
* SSLv3, TLS handshake, Server hello (2):
* SSLv3, TLS handshake, Certificate (11):
* SSLv3, TLS handshake, Server key exchange (12):
* SSLv3, TLS handshake, Request CERT (13):
* SSLv3, TLS handshake, Server finished (14):
* SSLv3, TLS handshake, Certificate (11):
* SSLv3, TLS handshake, Client key exchange (16):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* Server certificate:
*        subject: CN=icinga.<redacted>.de
*        start date: 2018-02-27 13:40:15 GMT
*        expire date: 2033-02-23 13:40:15 GMT
*        subjectAltName: icinga.<redacted>.de matched
*        issuer: CN=Icinga CA
*        SSL certificate verify ok.
* Server auth using Basic with user 'client-reschedule_check'
> POST /v1/actions/reschedule-check?filter=host.name==%22<redacted>%22&&type=Service HTTP/1.1
> Authorization: Basic <redacted>
> User-Agent: curl/7.37.0
> Host: icinga.<redacted>.de:5665
> Accept: application/json
> Content-Length: 23
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 23 out of 23 bytes
< HTTP/1.1 200 OK
< Content-Type: application/json
< Transfer-Encoding: chunked
* Server Icinga/r2.10.1-1 is not blacklisted
< Server: Icinga/r2.10.1-1
<
{"results":[{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!cpu-utiliziation'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!x11-server-is-running'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!acceptable-number-of-zombie-processes'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!disk-utilization'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!accpetable-number-of-user-sessions'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!raj-upgrade-works'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!acceptable-number-of-processes'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!accepts-ssh-login'."},{"code":200.0,"status":"Successf* Connection #0 to host icinga.<redacted>.de left intact
ully rescheduled check for object '<redacted>!gssd-health'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!mail-queue-length-normal'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!configs-up-to-date'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!OS-is-supported'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!Icinga-Client'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!irb-profile-adapt-succeeded'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!ntp-synchronizes-system-time'."},{"code":200.0,"status":"Successfully rescheduled check for object '<redacted>!system-is-up-to-date'."}]}
dns_resolution: 0,004, tcp_established: 0,005, ssl_handshake_done: 132,316, TTFB: 132,353

real    2m12.364s
user    0m0.022s
sys     0m0.009s
```

Still sometimes the command is executed faster.
```
real    0m0.055s
user    0m0.008s
sys     0m0.000s
```

In the Icinga logs are the following entrys which could be related.
```
[2018-10-19 19:29:10 +0200] information/WorkQueue: #6 (ApiListener, RelayQueue) items: 0, rate: 35.0667/s (2104/min 10736/5min 31735/15min);
[2018-10-19 19:29:19 +0200] information/WorkQueue: #7 (ApiListener, SyncQueue) items: 0, rate:  0/s (0/min 0/5min 1/15min);
[2018-10-19 19:29:23 +0200] information/WorkQueue: #13 (JsonRpcConnection, #2) items: 0, rate: 4.3/s (258/min 1252/5min 3756/15min);
[2018-10-19 19:29:23 +0200] information/WorkQueue: #12 (JsonRpcConnection, #1) items: 0, rate: 4.53333/s (272/min 1430/5min 4260/15min);
[2018-10-19 19:29:23 +0200] information/WorkQueue: #17 (JsonRpcConnection, #6) items: 0, rate: 4.16667/s (250/min 1278/5min 3821/15min);
[2018-10-19 19:29:23 +0200] information/WorkQueue: #11 (JsonRpcConnection, #0) items: 0, rate: 4.36667/s (262/min 1336/5min 3976/15min);
[2018-10-19 19:29:23 +0200] information/WorkQueue: #16 (JsonRpcConnection, #5) items: 0, rate: 4.43333/s (266/min 1368/5min 3944/15min);
[2018-10-19 19:29:23 +0200] information/WorkQueue: #18 (JsonRpcConnection, #7) items: 0, rate: 3.83333/s (230/min 1220/5min 3658/15min);
[2018-10-19 19:29:23 +0200] information/WorkQueue: #14 (JsonRpcConnection, #3) items: 0, rate:  4/s (240/min 1216/5min 3660/15min);
[2018-10-19 19:29:23 +0200] information/WorkQueue: #15 (JsonRpcConnection, #4) items: 0, rate: 4.26667/s (256/min 1276/5min 3904/15min);

[2018-10-19 19:34:20 +0200] information/WorkQueue: #6 (ApiListener, RelayQueue) items: 0, rate: 36.5833/s (2195/min 10476/5min 31524/15min);
```

Could it be related to the "max_anonymous_clients" in the API-Listener?
[https://icinga.com/docs/icinga2/latest/doc/09-object-types/#apilistener](url)

## Your Environment

* Version used (`icinga2 --version`):  
```
icinga2 - The Icinga 2 network monitoring daemon (version: r2.10.1-1)

Copyright (c) 2012-2018 Icinga Development Team (https://icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

System information:
  Platform: Debian GNU/Linux
  Platform version: 9 (stretch)
  Kernel: Linux
  Kernel version: 4.9.0-8-amd64
  Architecture: x86_64

Build information:
  Compiler: GNU 6.3.0
  Build host: 1880c6977367

Application information:

General paths:
  Config directory: /etc/icinga2
  Data directory: /var/lib/icinga2
  Log directory: /var/log/icinga2
  Cache directory: /var/cache/icinga2
  Spool directory: /var/spool/icinga2
  Run directory: /run/icinga2

Old paths (deprecated):
  Installation root: /usr
  Sysconf directory: /etc
  Run directory (base): /run
  Local state directory: /var

Internal paths:
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid
```
* Operating System and version:
```
Debian GNU/Linux 9 (stretch)
```
* Enabled features (`icinga2 feature list`):
``` 
Disabled features: command compatlog debuglog elasticsearch gelf graphite livestatus opentsdb perfdata statusdata syslog
Enabled features: api checker ido-mysql influxdb mainlog notification
```

**Comment by Allandor** [comment]
The node Setup is also affected.
```
icinga2 node setup --ticket  <redacted> \
> --endpoint icinga.<redacted>.de \
> --zone $(hostname).$(dnsdomainname) \
> --parent_host icinga.<redacted>.de \
> --trustedcert /var/lib/icinga2/certs/trusted-master.crt \
> --accept-commands --accept-config --disable-confd
information/cli: Requesting certificate with ticket '<redacted>'.
information/cli: Verifying parent host connection information: host 'icinga.<redacted>.de', port '5665'.
information/cli: Using the following CN (defaults to FQDN): '<redacted>'.
information/cli: Backup file '/var/lib/icinga2/certs//<redacted>.de.key.orig' already exists. Skipping backup.
information/cli: Backup file '/var/lib/icinga2/certs//<redacted>.de.crt.orig' already exists. Skipping backup.
information/base: Writing private key to '/var/lib/icinga2/certs//<redacted>.de.key'.
information/base: Writing X509 certificate to '/var/lib/icinga2/certs//<redacted>.de.crt'.
information/cli: Verifying trusted certificate file '/var/lib/icinga2/certs/trusted-master.crt'.
information/cli: Requesting a signed certificate from the parent Icinga node.
critical/cli: Client TLS handshake failed: Error: Timeout during handshake.

critical/cli: Failed to fetch signed certificate from parent Icinga node 'icinga.<redacted>.de, 5665'. Please try again. 
```

I will try to create a script to reproduce the behaviour.

**Comment by dnsmichi** [comment]
I suspect that the fix for #6635 affects this behaviour. Can you test the snapshot packages once built?

**Comment by Allandor** [comment]
It seems that one script used wrong credentials and these calls slow the API. With the snapshot package it works most of the time. I will keep on testing.

**Comment by dnsmichi** [comment]
Thanks. From our investigation, this doesn't happen anymore and thus being fixed with the coming 2.10.2 release.

**Comment by solikeumyeah** [comment]
To help desperate passers-by: Updating php may result in this error. Just restart Apache.
